File: plugins/scanners/csharp/dotnet/HttpEndpoints.cs
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Syntax;

namespace Groma.CSharpScanner;

/// <summary>
/// Endpoints ASP.NET Core serves: attribute-routed controller actions and minimal API route builders, including
/// route groups. A route that is not literal, or a builder that is not proven to be the application or a group,
/// reports nothing rather than a guessed path.
/// </summary>
internal static class HttpEndpoints
{
    private static readonly Dictionary<string, string> ActionVerbs = new(StringComparer.Ordinal)
    {
        ["HttpGet"] = "GET", ["HttpPost"] = "POST", ["HttpPut"] = "PUT", ["HttpDelete"] = "DELETE",
        ["HttpPatch"] = "PATCH", ["HttpHead"] = "HEAD", ["HttpOptions"] = "OPTIONS",
    };

    private static readonly Dictionary<string, string> MapVerbs = new(StringComparer.Ordinal)
    {
        ["MapGet"] = "GET", ["MapPost"] = "POST", ["MapPut"] = "PUT", ["MapDelete"] = "DELETE", ["MapPatch"] = "PATCH",
    };

    /// <summary>
    /// Settings anywhere in the scanned source that change what [controller] and [action] stand for: assigning
    /// SuppressAsyncSuffixInActionNames anything but true keeps Async in action names, and a
    /// RouteTokenTransformerConvention rewrites every token value.
    /// </summary>
    internal sealed record TokenConventions(bool KeepsAsyncSuffix, bool TransformsTokens);

    public static async Task<TokenConventions> Conventions(IEnumerable<Document> documents, CancellationToken cancellationToken)
    {
        bool keepsAsyncSuffix = false;
        bool transformsTokens = false;
        foreach (Document document in documents)
        {
            if (await document.GetSyntaxRootAsync(cancellationToken) is not SyntaxNode root) continue;
            keepsAsyncSuffix |= root.DescendantNodes().OfType<AssignmentExpressionSyntax>().Any(assignment =>
                HttpSyntax.CallName(assignment.Left) == "SuppressAsyncSuffixInActionNames" && !assignment.Right.IsKind(SyntaxKind.TrueLiteralExpression));
            transformsTokens |= root.DescendantNodes().OfType<SimpleNameSyntax>()
                .Any(name => name.Identifier.ValueText == "RouteTokenTransformerConvention");
        }
        return new TokenConventions(keepsAsyncSuffix, transformsTokens);
    }

    public static List<ScanHttpEndpoint> Of(SyntaxNode root, SemanticModel model, string file, string repositoryRoot,
        TokenConventions conventions, CancellationToken cancellationToken)
    {
        List<ScanHttpEndpoint> endpoints = [];
        foreach (TypeDeclarationSyntax type in root.DescendantNodes().OfType<TypeDeclarationSyntax>())
        {
            cancellationToken.ThrowIfCancellationRequested();
            // A declarative client shares these attributes but sends requests, so only classes serve.
            if (type is not InterfaceDeclarationSyntax) Controller(type, model, file, conventions, endpoints, cancellationToken);
        }
        foreach (InvocationExpressionSyntax call in root.DescendantNodes().OfType<InvocationExpressionSyntax>())
        {
            cancellationToken.ThrowIfCancellationRequested();
            Mapped(call, model, file, repositoryRoot, endpoints, cancellationToken);
        }
        return endpoints;
    }

    private static void Controller(TypeDeclarationSyntax type, SemanticModel model, string file, TokenConventions conventions,
        List<ScanHttpEndpoint> endpoints, CancellationToken cancellationToken)
    {
        if (ControllerRoutes(type, model, cancellationToken) is not { } routes || ClassPrefixes(routes, cancellationToken) is not { } prefixes) return;
        string? controller = ControllerName(type, conventions);
        foreach (MethodDeclarationSyntax action in type.Members.OfType<MethodDeclarationSyntax>())
        {
            if (!IsAction(action, model, cancellationToken) || Selectors(action, model, cancellationToken) is not { } selectors) continue;
            string operation = OperationId.Of(file, action);
            string? actionName = ActionName(action, model, conventions, cancellationToken);
            foreach ((string method, string template) in Routes(prefixes, selectors))
            {
                if (Tokens(template, controller, actionName) is string path && HttpRoutes.Endpoint(path) is ScanHttpSegment[] segments)
                    endpoints.Add(new ScanHttpEndpoint(operation, method, segments));
            }
        }
    }

    /// <summary>Each method and template an action serves: every route it declares under every class prefix.</summary>
    private static IEnumerable<(string Method, string Template)> Routes(List<string> prefixes, List<Selector> selectors)
    {
        foreach (string prefix in prefixes.Count == 0 ? [""] : prefixes)
        {
            // Without a class prefix, an action needs its own template to be routable.
            foreach (Selector selector in selectors.Where(selector => prefix.Length > 0 || selector.Template.Length > 0))
            {
                foreach (string method in selector.Methods) yield return (method, HttpRoutes.Join(prefix, selector.Template));
            }
        }
    }

    /// <summary>
    /// ASP.NET Core's controller rule as far as source shows it: a public, top-level, non-generic, non-abstract class
    /// declared in one place, named *Controller, marked [Controller] or deriving from ControllerBase or Controller, and
    /// not [NonController], counting attributes its bases carry. A controller also serves its bases' public methods, so a
    /// base between it and ControllerBase that declares one, or that the source does not declare, leaves its routes
    /// unknown. Its class routes are the [Route] attributes of the most derived class that declares any, as MVC reads
    /// them, each with the semantic model of the file declaring it. Null when the type is not such a controller.
    /// </summary>
    private static List<(AttributeSyntax Attribute, SemanticModel Model)>? ControllerRoutes(TypeDeclarationSyntax type, SemanticModel model, CancellationToken cancellationToken)
    {
        if (model.GetDeclaredSymbol(type, cancellationToken) is not INamedTypeSymbol
            {
                TypeKind: TypeKind.Class, DeclaredAccessibility: Accessibility.Public, IsAbstract: false, IsGenericType: false,
                ContainingType: null, DeclaringSyntaxReferences.Length: 1,
            } symbol) return null;
        List<string> attributes = [.. Attributes(type)];
        List<(AttributeSyntax Attribute, SemanticModel Model)> routes = [.. RouteAttributes(type).Select(attribute => (attribute, model))];
        bool framework = false;
        for (INamedTypeSymbol? inherited = symbol.BaseType; inherited is { SpecialType: not SpecialType.System_Object }; inherited = inherited.BaseType)
        {
            if (inherited.Name is "ControllerBase" or "Controller")
            {
                framework = true;
                break;
            }
            if (inherited.DeclaringSyntaxReferences.IsEmpty || inherited.GetMembers().OfType<IMethodSymbol>()
                .Any(method => method is { MethodKind: MethodKind.Ordinary, DeclaredAccessibility: Accessibility.Public, IsStatic: false })) return null;
            TypeDeclarationSyntax[] parts = [.. inherited.DeclaringSyntaxReferences.Select(part => part.GetSyntax(cancellationToken)).OfType<TypeDeclarationSyntax>()];
            attributes.AddRange(parts.SelectMany(Attributes));
            if (routes.Count == 0)
                routes.AddRange(parts.SelectMany(part => RouteAttributes(part).Select(attribute => (attribute, ModelOf(part.SyntaxTree, model.Compilation)))));
        }
        if (attributes.Contains("NonController")) return null;
        return framework || attributes.Contains("Controller") || symbol.Name.EndsWith("Controller", StringComparison.OrdinalIgnoreCase) ? routes : null;
    }

    private static IEnumerable<AttributeSyntax> RouteAttributes(TypeDeclarationSyntax type) =>
        type.AttributeLists.SelectMany(list => list.Attributes).Where(attribute => HttpSyntax.AttributeName(attribute) == "Route");

    /// <summary>A base may be declared in a referenced project, whose own compilation binds its constants.</summary>
    private static SemanticModel ModelOf(SyntaxTree tree, Compilation compilation) => compilation.ContainsSyntaxTree(tree)
        ? compilation.GetSemanticModel(tree)
        : compilation.References.OfType<CompilationReference>().Select(reference => reference.Compilation)
            .First(referenced => referenced.ContainsSyntaxTree(tree)).GetSemanticModel(tree);

    private static IEnumerable<string> Attributes(TypeDeclarationSyntax type) =>
        type.AttributeLists.SelectMany(list => list.Attributes).Select(HttpSyntax.AttributeName);

    /// <summary>ASP.NET Core's action rule: a public instance method with a body, not generic and not [NonAction].</summary>
    private static bool IsAction(MethodDeclarationSyntax action, SemanticModel model, CancellationToken cancellationToken) =>
        (action.Body is not null || action.ExpressionBody is not null)
        && model.GetDeclaredSymbol(action, cancellationToken) is IMethodSymbol
            { DeclaredAccessibility: Accessibility.Public, IsStatic: false, IsGenericMethod: false }
        && !action.AttributeLists.SelectMany(list => list.Attributes).Any(attribute => HttpSyntax.AttributeName(attribute) == "NonAction");

    /// <summary>
    /// The name [controller] stands for: the class name without its Controller suffix. Null when a token transformer
    /// rewrites it.
    /// </summary>
    private static string? ControllerName(TypeDeclarationSyntax type, TokenConventions conventions)
    {
        if (conventions.TransformsTokens) return null;
        string name = type.Identifier.ValueText;
        return name.EndsWith("Controller", StringComparison.OrdinalIgnoreCase) ? name[..^"Controller".Length] : name;
    }

    /// <summary>
    /// The name [action] stands for: a literal [ActionName], or else the method name without a trailing Async, which
    /// ASP.NET Core drops by default. Null when a token transformer rewrites it, when [ActionName] is not a constant
    /// the compilation proves, or when the source may keep the suffix and the name ends with Async.
    /// </summary>
    private static string? ActionName(MethodDeclarationSyntax action, SemanticModel model, TokenConventions conventions, CancellationToken cancellationToken)
    {
        if (conventions.TransformsTokens) return null;
        AttributeSyntax? named = action.AttributeLists.SelectMany(list => list.Attributes)
            .FirstOrDefault(attribute => HttpSyntax.AttributeName(attribute) == "ActionName");
        if (named is not null) return HttpSyntax.Constant(model, HttpSyntax.Argument(named), cancellationToken);
        string name = action.Identifier.ValueText;
        if (!name.EndsWith("Async", StringComparison.Ordinal)) return name;
        return conventions.KeepsAsyncSuffix ? null : name[..^"Async".Length];
    }

    /// <summary>
    /// Class-level route prefixes, rooted like an action template, or null when one is computed and hides every
    /// action's path.
    /// </summary>
    private static List<string>? ClassPrefixes(List<(AttributeSyntax Attribute, SemanticModel Model)> routes, CancellationToken cancellationToken)
    {
        List<string> prefixes = [];
        foreach ((AttributeSyntax attribute, SemanticModel model) in routes)
        {
            if (HttpSyntax.Constant(model, HttpSyntax.Argument(attribute), cancellationToken) is not string text) return null;
            prefixes.Add(HttpRoutes.Join("", text));
        }
        return prefixes;
    }

    /// <summary>One route of an action: the methods it allows, * for every method, on a template under the class prefix.</summary>
    private sealed record Selector(IReadOnlyList<string> Methods, string Template);

    /// <summary>
    /// The routes ASP.NET Core builds from an action's attributes. Each [Route] and each verb attribute with a
    /// template, Name or Order is one route: a verb attribute's route allows only its method, while a [Route] takes the
    /// methods of the template-less verb attributes. Those form one more route on the class prefix unless a [Route]
    /// took them. Null when a template is computed or [AcceptVerbs] states methods the scanner does not read.
    /// </summary>
    private static List<Selector>? Selectors(MethodDeclarationSyntax action, SemanticModel model, CancellationToken cancellationToken)
    {
        AttributeSyntax[] attributes = [.. action.AttributeLists.SelectMany(list => list.Attributes)];
        if (attributes.Any(attribute => HttpSyntax.AttributeName(attribute) == "AcceptVerbs")) return null;
        List<(string? Method, string Template)> routes = [];
        List<string> templatelessMethods = [];
        foreach (AttributeSyntax attribute in attributes.Where(IsRouting))
        {
            string? method = ActionVerbs.GetValueOrDefault(HttpSyntax.AttributeName(attribute));
            if (method is not null && HttpSyntax.Argument(attribute) is null && !DefinesRoute(attribute)) templatelessMethods.Add(method);
            else if (Template(attribute, model, cancellationToken) is string template) routes.Add((method, template));
            else return null;
        }
        List<string> routeMethods = templatelessMethods.Count > 0 ? templatelessMethods : ["*"];
        List<Selector> selectors = [.. routes.Select(route => new Selector(route.Method is null ? routeMethods : [route.Method], route.Template))];
        if (templatelessMethods.Count > 0 && routes.All(route => route.Method is not null)) selectors.Add(new Selector(templatelessMethods, ""));
        return selectors;
    }

    private static bool IsRouting(AttributeSyntax attribute) =>
        HttpSyntax.AttributeName(attribute) is var name && (name == "Route" || ActionVerbs.ContainsKey(name));

    /// <summary>A routing attribute's template: empty when it states none, null when it is computed.</summary>
    private static string? Template(AttributeSyntax attribute, SemanticModel model, CancellationToken cancellationToken) =>
        HttpSyntax.Argument(attribute) is ExpressionSyntax template ? HttpSyntax.Constant(model, template, cancellationToken) : "";

    /// <summary>A Name or Order makes a verb attribute define its own route even without a template.</summary>
    private static bool DefinesRoute(AttributeSyntax attribute) => attribute.ArgumentList?.Arguments
        .Any(argument => argument.NameEquals?.Name.Identifier.ValueText is "Name" or "Order") == true;

    /// <summary>
    /// Replaces the route tokens ASP.NET Core fills in. Any other token, or a token whose value the source does not
    /// prove, leaves the path unknown.
    /// </summary>
    private static string? Tokens(string path, string? controller, string? action)
    {
        if (controller is not null) path = path.Replace("[controller]", controller, StringComparison.OrdinalIgnoreCase);
        if (action is not null) path = path.Replace("[action]", action, StringComparison.OrdinalIgnoreCase);
        return path.Contains('[') || path.Contains(']') ? null : path;
    }

    private static void Mapped(InvocationExpressionSyntax call, SemanticModel model, string file, string repositoryRoot,
        List<ScanHttpEndpoint> endpoints, CancellationToken cancellationToken)
    {
        string name = HttpSyntax.CallName(call.Expression);
        bool listed = name == "MapMethods";
        if (!MapVerbs.TryGetValue(name, out string? verb) && !listed) return;
        SeparatedSyntaxList<ArgumentSyntax> arguments = call.ArgumentList.Arguments;
        if (arguments.Count < (listed ? 3 : 2)) return;
        if (HttpSyntax.Constant(model, arguments[0].Expression, cancellationToken) is not string pattern) return;
        if (Prefix(HttpSyntax.Receiver(call), model, 0, cancellationToken) is not string prefix) return;
        string? operation = Handler(arguments[listed ? 2 : 1].Expression, model, file, repositoryRoot, cancellationToken);
        if (operation is null || HttpRoutes.Endpoint(HttpRoutes.Under(prefix, pattern)) is not ScanHttpSegment[] segments) return;
        foreach (string method in listed ? Methods(model, arguments[1].Expression, cancellationToken) : [verb!])
            endpoints.Add(new ScanHttpEndpoint(operation, method, segments));
    }

    /// <summary>The operation that answers: the handler lambda, or the method a method group names.</summary>
    private static string? Handler(ExpressionSyntax handler, SemanticModel model, string file, string repositoryRoot, CancellationToken cancellationToken)
    {
        if (handler is AnonymousFunctionExpressionSyntax) return OperationId.Of(file, handler);
        SymbolInfo info = model.GetSymbolInfo(handler, cancellationToken);
        // An unresolved delegate type leaves the named method as the only candidate.
        IMethodSymbol? method = info.Symbol as IMethodSymbol
            ?? (info.CandidateSymbols.Length == 1 ? info.CandidateSymbols[0] as IMethodSymbol : null);
        return OperationId.OfMethod(repositoryRoot, method);
    }

    /// <summary>The literal methods of MapMethods, or none when any is computed.</summary>
    private static List<string> Methods(SemanticModel model, ExpressionSyntax expression, CancellationToken cancellationToken)
    {
        IEnumerable<ExpressionSyntax> elements = expression switch
        {
            ImplicitArrayCreationExpressionSyntax array => array.Initializer.Expressions,
            ArrayCreationExpressionSyntax array => array.Initializer?.Expressions ?? [],
            CollectionExpressionSyntax collection => collection.Elements.OfType<ExpressionElementSyntax>().Select(element => element.Expression),
            _ => [],
        };
        List<string> methods = [];
        foreach (ExpressionSyntax element in elements)
        {
            if (HttpSyntax.Constant(model, element, cancellationToken) is not string method) return [];
            methods.Add(method.ToUpperInvariant());
        }
        return methods;
    }

    /// <summary>
    /// The route prefix a builder expression carries: empty for the application itself, a group's literal prefixes,
    /// or null when the builder cannot be traced, such as a group passed in as a parameter.
    /// </summary>
    private static string? Prefix(ExpressionSyntax? builder, SemanticModel model, int depth, CancellationToken cancellationToken)
    {
        if (builder is null || depth > 8) return null;
        if (builder is InvocationExpressionSyntax call)
        {
            string name = HttpSyntax.CallName(call.Expression);
            // Building or creating the application starts at its root; other fluent calls keep the same builder.
            if (name == "Build" || (name == "Create" && HttpSyntax.Receiver(call) is ExpressionSyntax receiver
                && HttpSyntax.CallName(receiver) == "WebApplication")) return "";
            if (name != "MapGroup") return Prefix(HttpSyntax.Receiver(call), model, depth + 1, cancellationToken);
            string? parent = Prefix(HttpSyntax.Receiver(call), model, depth + 1, cancellationToken);
            string? pattern = call.ArgumentList.Arguments.Count == 0 ? null
                : HttpSyntax.Constant(model, call.ArgumentList.Arguments[0].Expression, cancellationToken);
            return parent is null || pattern is null ? null : HttpRoutes.Under(parent, pattern);
        }
        if (builder is MemberAccessExpressionSyntax access) return Prefix(access.Name, model, depth + 1, cancellationToken);
        if (builder is not SimpleNameSyntax) return null;
        // A name that holds one value carries that value's prefix; any name declared as a WebApplication is the root.
        if (HttpSyntax.SingleValue(model, builder, cancellationToken) is ExpressionSyntax value) return Prefix(value, model, depth + 1, cancellationToken);
        return HttpSyntax.DeclaredTypeName(model.GetSymbolInfo(builder, cancellationToken).Symbol) == "WebApplication" ? "" : null;
    }
}

File: plugins/scanners/javascript/src/http-endpoints.ts
import type { ScanHttpEndpoint } from '@groma/scanner'
import ts from 'typescript'
import { methodName } from '../../http-values.ts'
import { endpointPath } from './http-paths.ts'
import {
  functionValue, instanceValue, literalText, methodFromText, objectValue, propertyName, propertyValue,
  type HttpReader,
} from './http-reads.ts'
import { requiredModule } from './http-scope.ts'

// The supported routers follow the reference ../../typescript/src/http-endpoints.ts; change both together.
// This scanner reads one file at a time, so a router mounted in another file states no path and
// reports nothing, and only routes registered beside their application are supported.

/** Route registration members shared by Express, Fastify and Hono; `all` serves every method. */
const routeMembers = new Map([
  ['get', 'GET'], ['post', 'POST'], ['put', 'PUT'], ['patch', 'PATCH'],
  ['delete', 'DELETE'], ['head', 'HEAD'], ['options', 'OPTIONS'], ['all', '*'],
])

/** A router must be mounted to serve anything; an application instance serves from the root. */
type RegistrarKind = 'app' | 'router'

interface Registrar {
  kind: RegistrarKind
  /** The path this instance adds before any mount, or undefined when it states one the scanner cannot read. */
  own: string | undefined
}

type Registrars = Map<ts.Node, Registrar>

interface Mount {
  /** The instance this mount hangs under; its own prefixes precede this one. */
  parent: ts.Node
  /** Undefined when the prefix is not literal, which leaves the routes unsupported. */
  prefix?: string
}

export interface EndpointContext {
  reader: HttpReader
  /** The operation that runs this node: the enclosing function, else the module. */
  operationAt(node: ts.Node): string
}

/** Where a constructor comes from, whether this file imports the module or requires it. */
function constructorModule(reader: HttpReader, node: ts.Node): { module: string; name: string } | undefined {
  const required = requiredModule(node)
  if (required !== undefined) return { module: required, name: 'default' }
  return reader.scope.originOf(node)
}

/** A Koa router's own prefix option; an option the scanner cannot read leaves its path unknown. */
function routerPrefix(reader: HttpReader, initializer: ts.CallExpression | ts.NewExpression): string | undefined {
  const config = initializer.arguments?.[0]
  if (config === undefined) return ''
  const options = objectValue(reader, config)
  if (options === undefined) return undefined
  const prefix = propertyValue(options, 'prefix')
  return prefix === undefined ? '' : literalText(reader, prefix)
}

/** What a constructor from one module builds; only a router states a path of its own. */
function registrarFromModule(
  reader: HttpReader, origin: { module: string; name: string },
  initializer: ts.CallExpression | ts.NewExpression,
): Registrar | undefined {
  if (origin.module === 'express') return { kind: origin.name === 'Router' ? 'router' : 'app', own: '' }
  if (origin.module === 'fastify' || origin.module === 'koa') return { kind: 'app', own: '' }
  if (origin.module === 'hono' && origin.name === 'Hono') return { kind: 'app', own: '' }
  if (origin.module === '@koa/router' || origin.module === 'koa-router') {
    return { kind: 'router', own: routerPrefix(reader, initializer) }
  }
  return undefined
}

/** Recognize an instance by the module its constructor comes from, never by the members it offers. */
function registrarOf(reader: HttpReader, initializer: ts.Expression): Registrar | undefined {
  if (!ts.isCallExpression(initializer) && !ts.isNewExpression(initializer)) return undefined
  const callee = initializer.expression
  if (ts.isPropertyAccessExpression(callee)) {
    const holder = constructorModule(reader, callee.expression)
    return holder?.module === 'express' && callee.name.text === 'Router' ? { kind: 'router', own: '' } : undefined
  }
  const origin = constructorModule(reader, callee)
  return origin === undefined ? undefined : registrarFromModule(reader, origin, initializer)
}

function collectRegistrars(reader: HttpReader, source: ts.SourceFile): Registrars {
  const registrars: Registrars = new Map()
  const visit = (node: ts.Node): void => {
    if (ts.isVariableDeclaration(node) && node.initializer && ts.isIdentifier(node.name)
      && !reader.scope.reassigns(node.name.text)) {
      const registrar = registrarOf(reader, node.initializer)
      if (registrar !== undefined) registrars.set(node, registrar)
    }
    ts.forEachChild(node, visit)
  }
  ts.forEachChild(source, visit)
  return registrars
}

/**
 * `router.prefix('/api')` states the router's own path. A second prefix statement replaces the first
 * at runtime, and which one runs last is not a fact one file states, so the path becomes unknown.
 */
function applyPrefixCalls(reader: HttpReader, calls: readonly ts.CallExpression[], registrars: Registrars): void {
  for (const call of calls) {
    const callee = call.expression
    if (!ts.isPropertyAccessExpression(callee) || callee.name.text !== 'prefix') continue
    const declaration = receiverDeclaration(reader, call, registrars)
    const registrar = declaration === undefined ? undefined : registrars.get(declaration)
    if (declaration === undefined || registrar === undefined) continue
    const stated = literalText(reader, call.arguments[0])
    registrars.set(declaration, { ...registrar, own: registrar.own === '' ? stated : undefined })
  }
}

/** The registrar a call is made on, when the file declares it. */
function receiverDeclaration(reader: HttpReader, call: ts.CallExpression, registrars: Registrars): ts.Node | undefined {
  const receiver = ts.isPropertyAccessExpression(call.expression) ? call.expression.expression : undefined
  if (receiver === undefined || !ts.isIdentifier(receiver)) return undefined
  const declaration = reader.scope.declarationOf(receiver)
  return declaration !== undefined && registrars.has(declaration) ? declaration : undefined
}

/** The instance a mount argument names: the router itself, or the routes a Koa router exposes. */
function mountedRegistrar(reader: HttpReader, node: ts.Expression, registrars: Registrars): ts.Node | undefined {
  const named = ts.isCallExpression(node) && ts.isPropertyAccessExpression(node.expression)
    && node.expression.name.text === 'routes' ? node.expression.expression : node
  if (!ts.isIdentifier(named)) return undefined
  const declaration = reader.scope.declarationOf(named)
  return declaration !== undefined && registrars.has(declaration) ? declaration : undefined
}

/**
 * `use('/api', router)`, `route('/v1', child)` and `use(child.routes())` mount one instance. Only a
 * first argument the mounted instance does not occupy states a prefix.
 */
function mountEntries(
  reader: HttpReader, call: ts.CallExpression, registrars: Registrars,
): { child: ts.Node; mount: Mount }[] {
  const callee = call.expression
  if (!ts.isPropertyAccessExpression(callee)) return []
  const member = callee.name.text
  if (member !== 'use' && member !== 'route') return []
  // An unrecognized host, such as `createApp().use('/api', router)`, hides its own prefix.
  const parent = receiverDeclaration(reader, call, registrars)
  if (parent === undefined) return []
  const entries: { child: ts.Node; mount: Mount }[] = []
  for (const [index, argument] of call.arguments.entries()) {
    const child = mountedRegistrar(reader, argument, registrars)
    if (child === undefined) continue
    const prefix = index === 0 ? '' : literalText(reader, call.arguments[0])
    entries.push({ child, mount: { parent, ...(prefix === undefined ? {} : { prefix }) } })
  }
  return entries
}

function collectMounts(
  reader: HttpReader, calls: readonly ts.CallExpression[], registrars: Registrars,
): Map<ts.Node, Mount[]> {
  const mounts = new Map<ts.Node, Mount[]>()
  for (const call of calls) {
    for (const entry of mountEntries(reader, call, registrars)) {
      mounts.set(entry.child, [...mounts.get(entry.child) ?? [], entry.mount])
    }
  }
  return mounts
}

/** Where this instance hangs, before its own path: the root for an application, its mounts for a router. */
function mountBases(
  declaration: ts.Node, registrars: Registrars, mounts: Map<ts.Node, Mount[]>, seen: ReadonlySet<ts.Node>,
): string[] | undefined {
  const entries = mounts.get(declaration)
  if (entries === undefined || entries.length === 0) {
    return registrars.get(declaration)?.kind === 'router' ? undefined : ['']
  }
  const bases: string[] = []
  for (const entry of entries) {
    // Each mount is its own branch, so a second mount of the same instance is not a cycle.
    const parents = entry.prefix === undefined
      ? undefined : mountPrefixes(entry.parent, registrars, mounts, new Set([...seen, declaration]))
    if (parents === undefined) return undefined
    bases.push(...parents.map(parent => `${parent}/${entry.prefix}`))
  }
  return bases
}

/** Every path this instance serves under, or undefined when a mount or its own path is unknown. */
function mountPrefixes(
  declaration: ts.Node, registrars: Registrars, mounts: Map<ts.Node, Mount[]>, seen: ReadonlySet<ts.Node> = new Set(),
): string[] | undefined {
  const registrar = registrars.get(declaration)
  if (registrar === undefined || registrar.own === undefined || seen.has(declaration)) return undefined
  const bases = mountBases(declaration, registrars, mounts, seen)
  return bases === undefined ? undefined : [...new Set(bases.map(base => `${base}/${registrar.own}`))]
}

function endpointsFor(
  prefixes: readonly string[], route: string, methods: readonly string[], operation: string,
): ScanHttpEndpoint[] {
  const endpoints: ScanHttpEndpoint[] = []
  for (const prefix of prefixes) {
    const path = endpointPath(`${prefix}/${route}`)
    if (path === undefined) continue
    for (const method of methods) endpoints.push({ operation, method, path })
  }
  return endpoints
}

/** The operation a handler names, or the one that registers the route when the handler is unresolved. */
function handlerOperation(context: EndpointContext, handler: ts.Node | undefined, registration: ts.Node): string {
  const target = handler === undefined ? undefined : functionValue(context.reader, handler)
  return context.operationAt(target ?? registration)
}

function methodList(reader: HttpReader, node: ts.Node | undefined): string[] {
  if (node === undefined) return []
  const values = ts.isArrayLiteralExpression(node) ? [...node.elements] : [node]
  const methods = values.map(value => methodName(reader, value))
  return methods.every(method => method !== undefined) ? methods as string[] : []
}

/** Fastify's `route({ method, url, handler })` form. */
function routeOptions(
  context: EndpointContext, call: ts.CallExpression, prefixes: readonly string[],
): ScanHttpEndpoint[] {
  const options = objectValue(context.reader, call.arguments[0])
  if (options === undefined) return []
  const url = literalText(context.reader, propertyValue(options, 'url'))
  const methods = methodList(context.reader, propertyValue(options, 'method'))
  if (url === undefined || methods.length === 0) return []
  return endpointsFor(prefixes, url, methods, handlerOperation(context, propertyValue(options, 'handler'), call))
}

/**
 * The path a registration states. Some routers accept `get(name, path, handler)`, where the first
 * argument names the route, so a value that is not a path is never taken for one.
 */
function routePattern(reader: HttpReader, call: ts.CallExpression): string | undefined {
  const stated = literalText(reader, call.arguments[0])
  if (stated === undefined) return undefined
  if (stated.startsWith('/')) return stated
  const named = literalText(reader, call.arguments[1])
  return named?.startsWith('/') === true ? named : undefined
}

/** A registration needs its own handler: `app.get('name')` alone reads a setting. */
function routeEndpoints(
  context: EndpointContext, call: ts.CallExpression, registrars: Registrars, mounts: Map<ts.Node, Mount[]>,
): ScanHttpEndpoint[] {
  const callee = call.expression
  if (!ts.isPropertyAccessExpression(callee)) return []
  const declaration = receiverDeclaration(context.reader, call, registrars)
  if (declaration === undefined) return []
  const prefixes = mountPrefixes(declaration, registrars, mounts)
  if (prefixes === undefined) return []
  if (callee.name.text === 'route' && call.arguments.length === 1) return routeOptions(context, call, prefixes)
  const method = routeMembers.get(callee.name.text)
  if (method === undefined || call.arguments.length < 2) return []
  const route = routePattern(context.reader, call)
  if (route === undefined) return []
  const handler = call.arguments[call.arguments.length - 1]
  return endpointsFor(prefixes, route, [method], handlerOperation(context, handler, call))
}

/** `Bun.serve` on the runtime global, or the `serve` the `bun` module exports. */
function isBunServe(reader: HttpReader, call: ts.CallExpression): boolean {
  const callee = call.expression
  if (ts.isPropertyAccessExpression(callee)) {
    const receiver = callee.expression
    return ts.isIdentifier(receiver) && receiver.text === 'Bun' && callee.name.text === 'serve'
      && reader.scope.declarationOf(receiver) === undefined
  }
  return ts.isIdentifier(callee) && callee.text === 'serve' && reader.scope.originOf(callee)?.module === 'bun'
}

/** A route object's keys are methods, so an unknown key is not a handler. */
function serveMethod(key: string | undefined): string | undefined {
  const method = methodFromText(key)
  return method !== undefined && [...routeMembers.values()].includes(method) ? method : undefined
}

function methodHandlers(
  context: EndpointContext, route: string, byMethod: ts.ObjectLiteralExpression, call: ts.CallExpression,
): ScanHttpEndpoint[] {
  const endpoints: ScanHttpEndpoint[] = []
  for (const property of byMethod.properties) {
    if (!ts.isPropertyAssignment(property)) continue
    const method = serveMethod(propertyName(property))
    if (method === undefined) continue
    endpoints.push(...endpointsFor([''], route, [method], handlerOperation(context, property.initializer, call)))
  }
  return endpoints
}

/**
 * A route value is one handler for every method, or an object of handlers by method. An object the
 * scanner cannot read, such as one a spread fills, claims no method at all.
 */
function serveRoute(
  context: EndpointContext, route: string, value: ts.Expression, call: ts.CallExpression,
): ScanHttpEndpoint[] {
  const resolved = ts.isObjectLiteralExpression(value) ? value : instanceValue(context.reader, value)
  if (resolved !== undefined && ts.isObjectLiteralExpression(resolved)) {
    const byMethod = objectValue(context.reader, resolved)
    return byMethod === undefined ? [] : methodHandlers(context, route, byMethod, call)
  }
  return endpointsFor([''], route, ['*'], handlerOperation(context, value, call))
}

function serveEndpoints(context: EndpointContext, call: ts.CallExpression): ScanHttpEndpoint[] {
  if (!isBunServe(context.reader, call)) return []
  const options = objectValue(context.reader, call.arguments[0])
  const routes = options === undefined ? undefined : objectValue(context.reader, propertyValue(options, 'routes'))
  if (routes === undefined) return []
  const endpoints: ScanHttpEndpoint[] = []
  for (const property of routes.properties) {
    if (!ts.isPropertyAssignment(property)) continue
    const route = propertyName(property)
    if (route !== undefined) endpoints.push(...serveRoute(context, route, property.initializer, call))
  }
  return endpoints
}

/** Endpoints this file serves; an unsupported route, prefix or mount reports nothing. */
export function httpEndpoints(
  context: EndpointContext, source: ts.SourceFile, calls: readonly ts.CallExpression[],
): ScanHttpEndpoint[] {
  const registrars = collectRegistrars(context.reader, source)
  applyPrefixCalls(context.reader, calls, registrars)
  const mounts = collectMounts(context.reader, calls, registrars)
  const endpoints: ScanHttpEndpoint[] = []
  for (const call of calls) {
    endpoints.push(...routeEndpoints(context, call, registrars, mounts))
    endpoints.push(...serveEndpoints(context, call))
  }
  return endpoints
}
